src/main/scala/org/scalacheck/Gen.scala

```diff
@@ -467,6 +467,53 @@ object Gen extends GenArities{
     g.map(b.fromIterable)
   }
 
+  /** Monadic recursion on Gen
+   * This is a stack-safe loop that is the same as:
+   *
+   * {{{
+   *
+   * fn(a).flatMap {
+   *   case Left(a) => tailRec(a)(fn)
+   *   case Right(b) => Gen.const(b)
+   *   }
+   *
+   * }}}
+   *
+   * which is useful for doing monadic loops without blowing up the
+   * stack
+   */
+  def tailRecM[A, B](a0: A)(fn: A => Gen[Either[A, B]]): Gen[B] = {
+    @tailrec
+    def tailRecMR(a: A, seed: Seed, labs: Set[String])(fn: (A, Seed) => R[Either[A, B]]): R[B] = {
+      val re = fn(a, seed)
+      val nextLabs = labs | re.labels
+      re.retrieve match {
+        case None => r(None, re.seed).copy(l = nextLabs)
+        case Some(Right(b)) => r(Some(b), re.seed).copy(l = nextLabs)
+        case Some(Left(a)) => tailRecMR(a, re.seed, nextLabs)(fn)
+      }
+    }
+
+    // This is the "Reader-style" appoach to making a stack-safe loop:
+    // we put one outer closure around an explicitly tailrec loop
+    gen[B] { (p: P, seed: Seed) =>
+      tailRecMR(a0, seed, Set.empty) { (a, seed) => fn(a).doApply(p, seed) }
+    }
+  }
+
+  /** Build a pair from two generators
+   */
+  def product[A, B](ga: Gen[A], gb: Gen[B]): Gen[(A, B)] =
```

Comment:
The implementation is a bit different, but there is a [zip](https://github.com/rickynils/scalacheck/blob/35d5948a9185f509247b165ae9f02ba8c105c008/src/main/scala/org/scalacheck/GenArities.scala#L270) method. Is this intended to be available in addition to that?

Reply:
honestly, I didn't know about it. I think this implementation is more efficient but I am happy to remove it (or change the zip 2 implementation to use this approach).

My main interest in the tailRecM.

Reply:
@johnynek just to provide some context for that last comment: I'm not a Scalacheck maintainer; I just happened to see this change and I was pretty sure that a method similar to `product` existed, so I thought that I would point it out. I have no opinions on how to proceed regarding `product`, but I certainly could see benefit in `tailRecM` existing.

Reply:
Let's remove `product` here and devote the energy to porting that strategy over to the generated `zip` implementations. I agree with @johnynek that this approach is probably more efficient, so let's definitely do that.